formula/src/main/java/com/instacart/formula/internal/ActionManager.kt

```diff
@@ -12,151 +14,182 @@ internal class ActionManager(
     private val manager: FormulaManagerImpl<*, *, *>,
     private val inspector: Inspector?,
 ) {
+    private var running: SingleRequestMap<Any, DeferredAction<*>>? = null
+
+    // For scheduling start/terminate operations in post-evaluation phase
+    private var checkToStartActionList: MutableList<DeferredAction<*>>? = null
+    private var checkToRemoveActionList: MutableList<DeferredAction<*>>? = null
+
     /**
-     * Currently running actions
+     * Validation mode checks that during re-evaluation there were
+     * no changes such as new actions declared or existing actions
+     * removed.
      */
-    private var running: LinkedHashSet<DeferredAction<*>>? = null
+    private var validationFrame: ActionValidationFrame? = null
+
 
     /**
-     * Action list provided by [Evaluation.actions]
+     * Called by FormulaManagerImpl before evaluation that will be run as
+     * part of validation.
      */
-    private var actions: Set<DeferredAction<*>> = emptySet()
+    fun prepareValidationRun() {
```

Comment:
This validation frame is set before validation evaluation. We capture previous new and removed actions which we will compare after run is finished.